ArmPkg/Library/ArmFfaLib/ArmFfaCommon.c

```diff
@@ -0,0 +1,758 @@
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  *PartId = (FfaArgs.Arg2 >> ARM_FFA_DEST_EP_SHIFT) & ARM_FFA_PARTITION_ID_MASK;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Get spmc or spmd partition id.
+
+  @param [out]    SpmPartId      spmc/spmd partition id
+
+  @retval EFI_SUCCESS
+  @retval Others              Error
+
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibSpmIdGet (
+  OUT UINT16  *SpmPartId
+  )
+{
+  EFI_STATUS    Status;
+  ARM_FFA_ARGS  FfaArgs;
+
+  if (SpmPartId == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+
+  FfaArgs.Arg0 = ARM_FID_FFA_SPM_ID_GET;
+
+  ArmCallFfa (&FfaArgs);
+
+  if (IsFfaError (&FfaArgs)) {
+    Status = FfaStatusToEfiStatus (FfaArgs.Arg2);
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get partition id. Status: %r\n",
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  *SpmPartId = (FfaArgs.Arg2 >> ARM_FFA_DEST_EP_SHIFT) & ARM_FFA_PARTITION_ID_MASK;
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Get Partition info.
+  If This function is called to get partition descriptors
+  (Flags isn't set with FFA_PART_INFO_FLAG_TYPE_COUNT),
+  It should call ArmFfaLibRxRelease() to release RX buffer.
+
+  @param [in]   ServiceGuid    Service guid.
+  @param [in]   Flags          If this function called to get partition desc
+                               and get successfully,
+                               Caller should release RX buffer by calling
+                               ArmFfaLibRxRelease
+  @param [out]  Count          Number of partition or partition descriptor
+  @param [out]  Size           Size of Partition Info structure in Rx Buffer
+
+  @retval EFI_SUCCESS
+  @retval Others               Error
+**/
+EFI_STATUS
+EFIAPI
+ArmFfaLibPartitionInfoGet (
+  IN  EFI_GUID  *ServiceGuid,
+  IN  UINT32    Flags,
+  OUT UINT32    *Count,
+  OUT UINT32    *Size           OPTIONAL
+  )
+{
+  EFI_STATUS    Status;
+  ARM_FFA_ARGS  FfaArgs;
+  UINT64        Uuid[2];
+  UINT32        *SmcUuid;
+
+  if (Count == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((((Flags >> FFA_PART_INFO_FLAG_TYPE_SHIFT) & FFA_PART_INFO_FLAG_TYPE_MASK) !=
+       FFA_PART_INFO_FLAG_TYPE_COUNT) && (Size == NULL))
+  {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if (ServiceGuid != NULL) {
+    ConvertEfiGuidToUuid (ServiceGuid, Uuid);
+    SmcUuid = (UINT32 *) Uuid;
+  } else {
+    ZeroMem (SmcUuid, sizeof (EFI_GUID));
+  }
+
+  ZeroMem (&FfaArgs, sizeof (ARM_FFA_ARGS));
+
+  FfaArgs.Arg0 = ARM_FID_FFA_PARTITION_INFO_GET;
+  FfaArgs.Arg1 = SmcUuid[0];
+  FfaArgs.Arg2 = SmcUuid[1];
+  FfaArgs.Arg3 = SmcUuid[2];
+  FfaArgs.Arg4 = SmcUuid[3];
+  FfaArgs.Arg5 = Flags;
+
+  ArmCallFfa (&FfaArgs);
+
+  if (IsFfaError (&FfaArgs)) {
+    Status = FfaStatusToEfiStatus (FfaArgs.Arg2);
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get partition information of %g. Status: %r\n",
+      __func__,
+      (EFI_GUID *)SmcUuid,

```

Comment:
Recommend to print the incoming GUID. This one is the mangled version, which needs extra brain power to find out the original one :)